gui/wxpython/main_window/notebook.py

```diff
@@ -0,0 +1,153 @@
+ - notebook::MapNotebook
+
+(C) 2022 by the GRASS Development Team
+
+This program is free software under the GNU General Public License
+(>=v2). Read the file COPYING that comes with GRASS for details.
+
+@author Linda Kladivova <lindakladivova gmail.com>
+@author Anna Petrasova <kratochanna gmail.com>
+"""
+
+import os
+
+import wx
+import wx.lib.agw.aui as aui
+
+from core import globalvar
+from gui_core.wrap import SimpleTabArt
+
+
+class MapPageFrame(wx.Frame):
+    """Frame for independent map display window."""
+
+    def __init__(self, parent, mapdisplay, size, pos, title):
+        wx.Frame.__init__(self, parent=parent, size=size, pos=pos, title=title)
+        self.mapdisplay = mapdisplay
+        self.mapdisplay.Reparent(self)
+
+        self._layout()
+
+        # set system icon
+        self.SetIcon(
+            wx.Icon(
+                os.path.join(globalvar.ICONDIR, "grass_map.ico"), wx.BITMAP_TYPE_ICO
+            )
+        )
+
+        self.mapdisplay.onFocus.emit()
+        self.Bind(wx.EVT_CLOSE, self.OnClose)
+
+        self._show()
+
+    def _layout(self):
+        sizer = wx.BoxSizer(wx.VERTICAL)
+        sizer.Add(self.mapdisplay, proportion=1, flag=wx.EXPAND)
+        self.SetSizer(sizer)
+
+    def _show(self):
+        """Show frame and contained mapdisplay panel"""
+        self.mapdisplay.Show()
+        self.Show()
+
+    def SetDockingCallback(self, function):
+        """Set docking callback on reparented mapdisplay panel"""
+        self.mapdisplay.SetDockingCallback(function)
+
+    def OnClose(self, event):
+        """Close frame and associated layer notebook page."""
+        self.mapdisplay.OnCloseWindow(event=None, askIfSaveWorkspace=True)
+
+
+class MapNotebook(aui.AuiNotebook):
+    """Map notebook class. Overrides some AuiNotebook classes.
+    Takes into consideration the dock/undock functionality.
+    """
+
+    def __init__(
+        self,
+        parent,
+        agwStyle=aui.AUI_NB_DEFAULT_STYLE | aui.AUI_NB_TAB_EXTERNAL_MOVE | wx.NO_BORDER,
+    ):
+        self.parent = parent
+        super().__init__(parent=self.parent, id=wx.ID_ANY, agwStyle=agwStyle)
+
+        self.SetArtProvider(SimpleTabArt())
+
+        # bindings
+        self.Bind(
+            aui.EVT_AUINOTEBOOK_PAGE_CHANGED,
+            lambda evt: self.GetCurrentPage().onFocus.emit(),
+        )
+        self.Bind(aui.EVT_AUINOTEBOOK_PAGE_CLOSE, self.OnClose)
+
+    def UndockMapDisplay(self, page):
+        """Undock active map display to independent MapFrame object"""
+        index = self.GetPageIndex(page)
+        text = self.GetPageText(index)
+        original_size = page.GetSize()
+        original_pos = page.GetPosition()
+        self.RemovePage(index)
+        frame = MapPageFrame(
+            parent=self.parent,
+            mapdisplay=page,
+            size=original_size,
+            pos=original_pos,
+            title=text,
+        )
+        frame.SetDockingCallback(self.DockMapDisplay)
```

Comment:
This works for me:

```suggestion
        frame.Centre(wx.BOTH)
        frame.SetDockingCallback(self.DockMapDisplay)
```

Reply:
Thank you for testing @nilason! Just from my side, it might be better to put this line to the MapPageFrame init (all layout code is currently in _layout method). Is it still working after [60cd7da](https://github.com/OSGeo/grass/pull/2667/commits/60cd7da7ef39f36b542a1a9bec7047c26c61e7e5)?

Reply:
Good. I have used the CentreOnParent. For me there is no change in the results - from [1c4aef9](https://github.com/OSGeo/grass/pull/2667/commits/1c4aef99c22b1162ea89fe5c69001306c7d94b01) to [d5609a3](https://github.com/OSGeo/grass/pull/2667/commits/d5609a3af73d8f8ee79055b28a2f607bb093841a), I experienced the same behavior on Ubuntu.

Reply:
There should be no or little change for a maximized window.

Reply:
Thanks @lindakladivova for implementing and @nilason for testing. I'm resolving this conversation.